content/en/blog/2023/exponential-histograms.md

```diff
@@ -0,0 +1,196 @@
+exactly the base. For now, we will only consider nonnegative indices, but
+negative indexed buckets are also possible and define all buckets between 0
+and 1. Keeping with our example using a scale of 3 and resulting base of
+1.090508, the third bucket at index 2 has an upper bound of
+`1.090508^(2+1) = 1.29684`. The following table shows upper bounds for the first
+10 buckets of a few different scale factors:
+
+| index | scale -1 | scale 0 | scale 1 | scale 3 |
+| ----- | -------- | ------- | ------- | ------- |
+| -1    | **1**    | **1**   | **1**   | **1**   |
+| 0     | **4**    | 2       | 1.4142  | 1.0905  |
+| 1     | **16**   | **4**   | 2       | 1.1892  |
+| 2     | 64       | 8       | 2.8284  | 1.2968  |
+| 3     | 256      | **16**  | **4**   | 1.4142  |
+| 4     | 1024     | 32      | 5.6569  | 1.5422  |
+| 5     | 4096     | 64      | 8       | 1.6818  |
+| 6     | 16384    | 128     | 11.3137 | 1.8340  |
+| 7     | 65536    | 256     | **16**  | 2       |
+| 8     | 262144   | 512     | 22.6274 | 2.1810  |
+| 9     | 1048576  | 1024    | 32      | 2.3784  |
+
+I've bolded some of the values here to show an important property of exponential
+histograms called _perfect subsetting_.
+
+## Perfect subsetting
+
+In the chart above, some of the bucket boundaries are shared between histograms
+with differing scale factors. In fact, each time the scale factor increases by
+1, exactly 1 boundary is inserted between each existing boundary. This feature
+is called perfect subsetting because each set of boundaries for a given scale
+factor is a perfect subset of the boundaries for any histogram with a greater
+scale factor.
+
+Because of this, histograms with differing scale factors can be normalized to
+whichever has the lesser scale factor by combining neighboring buckets. This
+means that histograms with different scale factors can still be combined into a
+single histogram with exactly the precision of the least precise histogram being
+combined. For example, histogram _A_ with scale 3 and histogram _B_ with scale 2
+can be combined into a single histogram _C_ with scale 2 by first summing each
+pair of neighboring buckets in _A_ to form histogram _A'_ with scale 2. Then,
+each bucket in _A'_ is summed with the corresponding bucket of the same index in
+_B_ to make _C_.
+
+## Error rate
+
+The error rate of a histogram is defined as the average relative error when
+estimating the value of a particular-ranked data point. This is important
+because it is how φ-quantiles are estimated. In a histogram with x data points,
+the nth percentile is the data point at rank `n / 100 * x`. For an example of
+this calculation, see my previous post [Histograms vs Summaries][]. When
+estimating ɸ-quantiles it is very important to know the expected relative error
+rates and maximum relative error rates so that you can effectively monitor your
+SLOs.
+
+When using linear interpolation, the expected relative error is half the bucket
```

Comment:
The relative error is minimized, if the harmonic mean of the bucket boundaries is used for estimation.
For bucket with index `i`, the lower and upper boundaries are given by `l = base^i` and `u = base^(i + 1)`, respectively. The estimated value `x` given by the harmonic mean would be
```
x = 2*l*u / (l + u)
  = (2 * base^i * base^(i + 1)) / (base^i + base^(i + 1))
  = 2*base^(i+1) / (1 + base)
```
The worst case estimation error appears, if the original value was either close to the lower boundary `l` or the upper boundary `u`. We can calculate the relative error for both cases:
```
(x - l) / l = x/l -1 = (2*base) / (1 + base) - 1 = (base - 1) / (base + 1)
(u - x) / u = 1 - x/u = 1 - 2 / (1 + base) = (base - 1) / (base + 1)
```
Since the relative error is the same in both extreme cases, `(base - 1) / (base + 1)` also defines the maximum relative estimation error. For `scale = 3` with corresponding `base  = 1.090508` (see base calculation above) the maximum relative error would be 
```
(1.090508 - 1) / (1.090508 + 1) = 0.04329 = 4.329%
```